URL: http://dwm.suckless.org/patches/gapless_grid
     Add gapless grid layout.
NOTE: Not applicable to vanilla dwm.c

Index: dwm/config.h
==================================================================
--- /dev/null   1970-01-01 00:00:00.000000000 +0000
+++ dwm/config.h 2014-11-09 15:24:17.132117105 +0000

@@ -47,6 +47,7 @@ static const Layout layouts[] = {
{ "ý", NULL }, /* no layout function means floating behavior */
{ "ô", monocle },
{ "ù", spiral },
+ { "ú", gaplessgrid },

Index: dwm/dwm.c
===================================================================
--- /dev/null	1970-01-01 00:00:00.000000000 +0000
+++ dwm/dwm.c	2014-02-09 15:24:17.132117105 +0100

@@ -851,6 +911,44 @@ focusstack(const Arg *arg) {
        }   
 }
 
+void
+gaplessgrid(Monitor *m) {
+       unsigned int n, cols, rows, cn, rn, i, cx, cy, cw, ch;
+       Client *c;
+
+       for(n = 0, c = nexttiled(m->clients); c; c = nexttiled(c->next), n++) ;
+       if(n == 0)
+               return;
+
+       /* grid dimensions */
+       for(cols = 0; cols <= n/2; cols++)
+               if(cols*cols >= n)
+                       break;
+       if(n == 5) /* set layout against the general calculation: not 1:2:2, but 2:3 */
+               cols = 2;
+       rows = n/cols;
+
+       /* window geometries */
+       cw = cols ? m->ww / cols : m->ww;
+       cn = 0; /* current column number */
+       rn = 0; /* current row number */
+       for(i = 0, c = nexttiled(m->clients); c; i++, c = nexttiled(c->next)) {
+               if(i/rows + 1 > cols - n%cols)
+                       rows = n/cols + 1;
+               ch = rows ? m->wh / rows : m->wh;
+               cx = m->wx + cn*cw;
+               cy = m->wy + rn*ch;
+               resize(c, cx, cy, cw - 2 * c->bw, ch - 2 * c->bw, False);
+               rn++;
+               if(rn >= rows) {
+                       rn = 0;
+                       cn++;
+               }
+       }
+}
+
+
+

